Convert integration tests to Cypress

nemanjaglumac opened the pull request:
### Status
READY

### What does this PR accomplish?
- Converts remaining integration tests to Cypress test
- Removes all integration tests and their related files
    - There is one more related file left `frontend/test/__support__/integration.js` which is used by some unit test files. Once we remove them, it will be safe to remove it as well.
- Removes `fe-tests-integration` from CI

#### Context:
- Main list at https://github.com/metabase/metabase/issues/13657

### Additional notes and qeustions
- I have removed `AuditTable.integ.spec.jsx` in 1aa9aa225fc753cb8ef871a8154c75a00159c905. Although the file name suggests it should test audits, it was basically testing that there is pagination present when there are more than X rows.

There was this mock POST api call to generate the needed rows:
```javascript
mock.post("/api/dataset", (req, res) =>
  res.json({
    data: { rows: [[0]], cols: [{ name: "x" }] },
    row_count: 100,
    status: "completed",
  }),
);
```
My question is do we really need to migrate this to Cypress? If there is value in doing that, I'll follow up with another PR in order not to block this merge.

nemanjaglumac committed:
Run each test in isolation

nemanjaglumac committed:
Convert `EditUserModal.integ.spec.js` to Cypress

nemanjaglumac committed:
Convert `PeopleListingApp.integ.spec.js` to Cypress

nemanjaglumac committed:
Convert `GroupDetailApp.integ.spec.js` to Cypress

nemanjaglumac committed:
Delete `GroupsListingApp.integ.spec.js`

It was already covered in Cypress.

nemanjaglumac committed:
Delete `NewUserModal.integ.spec.js`

It was already covered in Cypress.

nemanjaglumac committed:
Convert `UserActivationModal.integ.spec.js`

nemanjaglumac committed:
Convert `UserPasswordResetModal.integ.spec.js`

nemanjaglumac committed:
Convert `UserSettings.integ.spec.js`

nemanjaglumac committed:
Remove integration tests

nemanjaglumac committed:
Remove integration tests from CI

nemanjaglumac committed:
Fix linter errors

nemanjaglumac committed:
Delete `AuditTable.integ.spec.jsx`

nemanjaglumac committed:
Clean up `people.cy.spec.js`

nemanjaglumac commented on docs/developers-guide.md:
````diff
@@ -120,7 +121,7 @@ All frontend tests are located in `frontend/test` directory. Run all frontend te
 ```
 
-which will run unit, integration and Cypress end-to-end tests in sequence.
+which will run unit, timezones and Cypress end-to-end tests in sequence.
````
Maybe I should've omitted `timezones`? After all, they are a form of unit tests if I'm not mistaken.

ariya replied:
Yes, it is a form of unit tests (although I'm not sure why there is a separate Jest config for that, though)

paulrosenzweig replied:
We run those with different pairs of client/reporting timezones. In the context of unit tests, the "client" timezone is node's timezone.

There have been bugs caused by the browser being east or west of the MB reporting timezone.

[Here's where we run those.](https://github.com/metabase/metabase/blob/master/frontend/test/__runner__/run_timezone_tests)

nemanjaglumac replied:
So, the final verdict is...? :)
Leave that piece of text or remove it? Not that it makes much difference, but I'm asking simply for the "correctness" sake.

paulrosenzweig replied:
Ah sorry I didn't comment on that. I might drop it, but either way is fine.

nemanjaglumac replied:
Dropped, done :)
Thanks

nemanjaglumac committed:
Update assertion that makes sure XHR didn't happen

nemanjaglumac commented on frontend/test/metabase/scenarios/setup/user_settings.cy.spec.js:
```diff
@@ -0,0 +1,41 @@
+    cy.findByText("Update").click();
+    cy.findByDisplayValue("John");
+
+    // It is hard and unreliable to assert that something didn't happen in Cypress
+    // This solution was the only one that worked out of all others proposed in this SO topic: https://stackoverflow.com/a/59302542/8815185
+    cy.get("@membership").then(() => {
+      expect(
+        cy.state("requests").filter(r => r.alias === "membership"),
+      ).to.have.length(0);
+    });
```
To test whether this really catches an XHR, I intentionally changed XHR to the one that is present. The test fails as expected:
![image](https://user-images.githubusercontent.com/31325167/103101002-b6cbd000-4615-11eb-8ba5-be5db2247062.png)

nemanjaglumac commented on frontend/test/metabase/scenarios/setup/user_settings.cy.spec.js:
```diff
@@ -0,0 +1,38 @@
+    cy.findByText("Update").click();
+    cy.findByDisplayValue("John");
+
+    // It is hard and unreliable to assert that something didn't happen in Cypress
+    // This solution was the only one that worked out of all others proposed in this SO topic: https://stackoverflow.com/a/59302542/8815185
+    cy.get("@membership").then(() => {
+      expect(requestsCount("membership")).to.have.length(0);
+    });
```
To test whether this really catches an XHR, I intentionally changed XHR **locally** to the one that is present. The test failed as expected:
![image](https://user-images.githubusercontent.com/31325167/103101286-2abaa800-4617-11eb-87f5-98dae165b78d.png)

ariya commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,10 +1,11 @@
 
--  [ ] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
-(unless it's a tiny documentation change).
+- [ ] Run the frontend and Cypress end-to-end tests with `yarn lint && yarn flow && yarn test`)
```
An opportunity to fix this: `yarn lint` already runs Flow (among others, such as ESLint, Prettier, etc).

nemanjaglumac replied:
Fixed it. Thank you for valuable suggestion.

ariya reviewed (approved):
Beside that nitpicky issue, it's a solid refactoring :+1:

nemanjaglumac committed:
Remove redundant command from PR template

Note: `yarn-lint` already runs `yarn flow`.
https://github.com/metabase/metabase/pull/14174#discussion_r549790198

nemanjaglumac committed:
Remove explicit mention of timezones in `developers-guide.md`
